pkg/common/plugin/httppop/httppop.go

```diff
@@ -0,0 +1,100 @@
+package httppop
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"io/ioutil"
+
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/common/agentpathtemplate"
+	"github.com/spiffe/spire/pkg/common/idutil"
+)
+
+const (
+	nonceLen = 32
+
+	// PluginName for http based attestor
+	PluginName = "httppop"
+)
+
+// DefaultAgentPathTemplate is the default template
+var DefaultAgentPathTemplate = agentpathtemplate.MustParse("/{{ .PluginName }}/{{ .HostName }}")
+
+type agentPathTemplateData struct {
+	HostName    string
+	PluginName  string
+	TrustDomain string
+}
+
+type AttestationData struct {
+	HostName  string `json:"hostname"`
+	AgentName string `json:"agentname"`
+	Port      int    `json:"port"`
+}
+
+type Challenge struct {
+	Nonce []byte `json:"nonce"`
+}
+
+type Response struct {
+}
+
+func GenerateChallenge() (*Challenge, error) {
+	nonce, err := generateNonce()
+	if err != nil {
+		return nil, err
+	}
+	return &Challenge{Nonce: nonce}, nil
+}
+
+func CalculateResponse(challenge *Challenge) (*Response, error) {
```

Comment:
What do you envision happening here? Since the challenge is fulfilled by out of band call back from the server, I don't think we need to send anything back on the stream?

Reply:
Mostly a copy/paste thing from other plugins that use it...

But, there does need to be a message from the agent to the server after the agent starts up the http server and hosts the token to tell the server it can now try and call it. So, there is a Response, even though its blank.